grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java

```diff
@@ -416,6 +428,52 @@ private static Descriptor findTypeDescriptor(FileDescriptor file, FieldDescripto
         return file.findMessageTypeByName(messageType.getName());
     }
 
+    // to make it more efficient, we calculate whether extract response body one time
+    // if there is no matching toplevel field, we set it to null
+    @Nullable
+    private static String calculateResponseBody(List<FieldDescriptor> topLevelFields, String responseBody) {
+        if (StringUtil.isNullOrEmpty(responseBody)) {
+            return null;
+        }
+        for (FieldDescriptor fieldDescriptor: topLevelFields) {
+            if (fieldDescriptor.getName().equals(responseBody)) {
+                return responseBody;
+            }
+        }
+        return null;
+    }
+
+    @Nullable
+    private static Function<HttpData, HttpData> generateResponseBodyConverter(TranscodingSpec spec) {
+        @Nullable final String responseBody = spec.responseBody;
+        if (responseBody == null) {
+            return null;
+        } else {
+            return httpData -> {
+                final byte[] array = httpData.array();
+                try {
+                    final JsonNode jsonNode = JacksonUtil.readValue(array, JsonNode.class);
+                    final String responseBodyJsonKey =
+                            CaseFormat.UPPER_UNDERSCORE.to(CaseFormat.LOWER_CAMEL, responseBody);
+                    final Iterator<String> fieldNames = jsonNode.fieldNames();
+                    while (fieldNames.hasNext()) {
+                        final String fieldName = fieldNames.next();
+                        if (CaseFormat.UPPER_UNDERSCORE.to(
+                                CaseFormat.LOWER_CAMEL, responseBody).equalsIgnoreCase(fieldName)) {
```

Comment:
Did we also want to compare with the original `responseBody` in case `preservingProtoFieldNames` is specified?

Reply:
Can we also add a test case for this behavior?

i.e.
```
.jsonMarshallerFactory(service -> GrpcJsonMarshaller
                                                               .builder()
                                                               .jsonMarshallerCustomizer(m -> m.preservingProtoFieldNames(true))
                                                               .build(service))
```

Reply:
Good point

Reply:
resolved